### PR TITLE
fix widget built twice during warm up frame

### PR DESCRIPTION
**Pull request by chunhtai**
## Description

The runapp work as follow
1, build rendering tree
2, schedule warm up frame
...(async gap)
3, run async draw frame

There is a chance engine will flush user setting in the middle of 2 and 3 which will dirty rendering tree. and draw frame will have to rebuild the rendering tree

With this pr the runapp work as follow
1, schedule build rendering tree
2, schedule warm up frame
...(async gap)
3, run async build rendering tree
4, run async draw frame

In this case, it does not matter whether something try to dirty render tree in the middle of 2 and 3 as there is no rendering tree to dirty.

This will however affect flutter test since it override schedulewarmup frame to be synchronized function to make sure runapp run synchronously. To addjust that, I implements a private runapp that flush timer after the regular runapp and removed the schedulewarmup override.

I do not use the callback solution we talked about @Hixie @goderbauer since that will create code duplication in flutter test binding and widgetflutterbinding. However, this approach might be a bigger breaking change for people who work directly with binding.

## Related Issues

https://github.com/flutter/flutter/issues/31195

## Tests

I added the following tests:

'WidgetBinding build rendering tree and warm up frame back to back'

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I signed the [CLA].
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [ ] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/widgets/binding.dart [hunk 1]
```diff
@@ -847,9 +847,11 @@ mixin WidgetsBinding on BindingBase, SchedulerBinding, GestureBinding, RendererB
 ///  * [WidgetsBinding.handleBeginFrame], which pumps the widget pipeline to
 ///    ensure the widget, element, and render trees are all built.
 void runApp(Widget app) {
-  WidgetsFlutterBinding.ensureInitialized()
-    ..attachRootWidget(app)
-    ..scheduleWarmUpFrame();
+  final WidgetsBinding binding = WidgetsFlutterBinding.ensureInitialized();
+  Timer.run(() {
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/binding.dart]: I might as well create an api to just schedule attachRootWidget and flutter test can override this api. This might make  it more clean, but I am not sure if it worth to expose an api for that

File: packages/flutter/test/widgets/nested_scroll_view_async_test.dart [hunk 2]
```diff
@@ -13,41 +13,41 @@ void main() {
...
+                  child: const Placeholder(),
                 ),
               ),
             ),
           ),
         ),
       );
-      // Marks element as dirty right before the first draw frame is called.
-      // This can happen when engine flush user setting.
+      // Initiates the first build.
```
Comment (chunhtai) [inline packages/flutter/test/widgets/nested_scroll_view_async_test.dart]: since it is no longer possible to dirty thing in the middle of warm up frame, I have to change the way to reproduce double builds. if we are ok to remove the null check in nestedscrollview, we can remove this test as well

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 3]
```diff
@@ -498,8 +498,6 @@ class _NestedScrollCoordinator implements ScrollActivityDelegate, ScrollHoldCont
 
   bool get hasScrolledBody {
     for (_NestedScrollPosition position in _innerPositions) {
-      // TODO(chunhtai): Replace null check with assert once
-      // https://github.com/flutter/flutter/issues/31195 is fixed.
       if (
         position.minScrollExtent != null &&
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart]: I intend to keep this check as there might be other corner case that hit this, but i cannot think of any now.

File: packages/flutter/test/widgets/binding_test.dart [hunk 4]
```diff
@@ -131,9 +131,5 @@ void main() {
     await tester.pump(const Duration(milliseconds: 1));
     expect(tester.binding.hasScheduledFrame, isFalse);
     expect(frameCount, 0);
-
-    tester.binding.scheduleWarmUpFrame(); // this actually tests flutter_test's implementation
-    expect(tester.binding.hasScheduledFrame, isFalse);
-    expect(frameCount, 1);
```
Comment (Hixie) [inline packages/flutter/test/widgets/binding_test.dart]: ?

File: packages/flutter/test/widgets/run_app_async_test.dart [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+          home: Material(
+            child: Text('test'),
+          ),
+        ),
+      );
+      // Rendering tree is not built synchronously.
+      expect(WidgetsBinding.instance.renderViewElement, isNull);
+      fakeAsync.flushTimers();
+      expect(WidgetsBinding.instance.renderViewElement, isNotNull);
+
```
Comment (Hixie) [inline packages/flutter/test/widgets/run_app_async_test.dart]: nit: extraneous blank line

File: packages/flutter_test/lib/src/binding.dart [hunk 6]
```diff
@@ -634,10 +634,14 @@ abstract class TestWidgetsFlutterBinding extends BindingBase
...
 
+  void _runAppSync(Widget app) {
+    runApp(app);
+  }
+
   Future<void> _runTestBody(Future<void> testBody(), VoidCallback invariantTester) async {
     assert(inTest);
 
-    runApp(Container(key: UniqueKey(), child: _preTestMessage)); // Reset the tree to a known state.
+    _runAppSync(Container(key: UniqueKey(), child: _preTestMessage)); // Reset the tree to a known state.
```
Comment (Hixie) [inline packages/flutter_test/lib/src/binding.dart]: Why this change? It seems identical?

Comment (Hixie) [reply packages/flutter_test/lib/src/binding.dart]: oh, i see, you override it later

File: packages/flutter_test/lib/src/binding.dart [hunk 7]
```diff
@@ -634,10 +634,14 @@ abstract class TestWidgetsFlutterBinding extends BindingBase
     return testCompleter.future;
   }
 
+  void _runAppSync(Widget app) {
```
Comment (Hixie) [inline packages/flutter_test/lib/src/binding.dart]: this shouldn't be private, and it should be documented (and @protected)

**Comment by Hixie** [comment]
This will definitely have to go through the breaking change announcement process. Might be worth trying to run it through the internal Google test suite and see how many things we break there, too, to get an idea of how much breakage this will cause.

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart (hunk 3 above)
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/nested_scroll_view.dart]: Turn these into asserts then?

File: packages/flutter/test/widgets/binding_test.dart (hunk 4 above)
Comment (chunhtai) [reply packages/flutter/test/widgets/binding_test.dart]: testerbinding is no longer overriding scheduleWarmUpFrame, there is no point to test it?

File: packages/flutter_test/lib/src/binding.dart (hunk 6 above)
Comment (chunhtai) [reply packages/flutter_test/lib/src/binding.dart]: This is really urgly, but i can't think of other way around it unless we add an api called scheduleAttachRootWidget, and override it in testbinding to make it synchronized

**Review by goderbauer — COMMENTED** [review]
Great debugging to figure this out!

Can you throw the "breaking change" label on here and sent out an announcement?

**Comment by goderbauer** [comment]
Looks like there are some tests on cirrus that were supposed to fail, but they are now passing.

**Review by goderbauer — COMMENTED** [review]
This looks good, but we should announce it as a breaking change and as Hixie pointed out, it may be good to run this through Google's internal test suite to see what breaks there.

**Comment by goderbauer** [comment]
@kpsroka FYI

**Comment by chunhtai** [comment]
internal test suite passed, breaking change email sent.

File: packages/flutter/lib/src/widgets/binding.dart [hunk 8]
```diff
@@ -787,6 +787,21 @@ mixin WidgetsBinding on BindingBase, SchedulerBinding, GestureBinding, RendererB
   Element get renderViewElement => _renderViewElement;
   Element _renderViewElement;
 
+  /// Schedules a [Timer] for attaching the root widget.
+  ///
+  /// This is called by [runApp] to configure the widget tree. Consider using
+  /// [attachRootWidget] if you want to build the widget tree synchronously.
+  ///
+  /// See also:
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/binding.dart]: nit: I would probably remove the see also here since the sentence above just said the same thing.

**Review by goderbauer — APPROVED** [review]
LGTM

Probably good to wait until next week with submission to give people some time to respond to the breaking change.

**Comment by tvolkert** [comment]
Was there a breaking change announcement for this change?  If so, @chunhtai will you please link it here?

**Comment by chunhtai** [comment]
> Was there a breaking change announcement for this change? If so, @chunhtai will you please link it here?

this is the one https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!searchin/flutter-announce/chun%7Csort:date/flutter-announce/m6ewfoO64mI/ziD4VmwUBQAJ

**Comment by tvolkert** [comment]
Thanks!
